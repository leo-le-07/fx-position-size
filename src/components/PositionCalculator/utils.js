// @flow

import axios from 'axios';

// https://1forge.com/forex-data-api
const FORGE_KEY = 'mFtJ9elrcPsjqexv1Aky0zdqsWsQs4jk';

export const calculateUsdPerPip = async (pair: string): Promise<?number> => {
  const response = await axios.get("https://forex.1forge.com/1.0.1/quotes?pairs="+pair+"&api_key="+FORGE_KEY);
  const { symbol, price } = response.data[0];
  if (symbol !== pair) return null;
  const currentPrice = parseFloat(price);
  return ((0.0001/currentPrice) * 100000) * currentPrice
};

