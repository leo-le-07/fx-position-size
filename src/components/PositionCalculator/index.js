// @flow

import React from 'react';
import Input from 'common/Input';
import { calculateUsdPerPip } from './utils.js';

type State = {
  accountSize: number,
  riskMoney: number,
  stopLossPips: number,
  pair: string,
  riskRatio: ?number,
  lots: ?number,
};

class PositionCalculator extends React.Component<{}, State> {
  state = {
    accountSize: 800,
    riskMoney: 20,
    stopLossPips: 10,
    pair: 'EURUSD',
    riskRatio: null,
    lots: null,
  }

  calculate = async () => {
    const { pair, riskMoney, stopLossPips, accountSize } = this.state;
    const usdPerPip = await calculateUsdPerPip(pair.toUpperCase());
    if (!usdPerPip) return this.setState({ lots: null, riskRatio:  null })
    const lotsResult = (riskMoney / (usdPerPip * stopLossPips));
    this.setState({
      lots: lotsResult,
      riskRatio: (riskMoney / accountSize) * 100
    })
  }

  handleOnChangeFor = (propertyName: string) => (event: SyntheticInputEvent<HTMLInputElement>) => {
    const value = event.target.value;
    this.setState({
      [propertyName]: value
    });
  };

  render() {
    const { accountSize, riskMoney, stopLossPips, pair, riskRatio, lots } = this.state;

    return (
      <div className="card">
        <div className="card-header">
          Position Size Calculator
        </div>
        <div className="card-body row">
          <div className="col-md-6">
            <h5 className="card-title">Inputs</h5>
            <div className="form-group row">
              <label htmlFor="account-size" className="col-md-4">Account Size (USD)</label>
              <div className="col-md-8">
                <Input
                  type="number"
                  id="account-size"
                  value={accountSize}
                  onChange={this.handleOnChangeFor('accountSize')}
                />
              </div>
            </div>
            <div className="form-group row">
              <label htmlFor="risk-money" className="col-md-4">Risk Money (USD)</label>
              <div className="col-md-8">
                <Input
                  type="number"
                  id="risk-money"
                  value={riskMoney}
                  onChange={this.handleOnChangeFor('riskMoney')}
                />
              </div>
            </div>
            <div className="form-group row">
              <label htmlFor="stop-loss-pips" className="col-md-4">Stop-Loss (pips)</label>
              <div className="col-md-8">
                <Input
                  type="number"
                  id="stop-loss-pips"
                  value={stopLossPips}
                  onChange={this.handleOnChangeFor('stopLossPips')}
                />
              </div>
            </div>
            <div className="form-group row">
              <label htmlFor="current-pair" className="col-md-4">Current Pair</label>
              <div className="col-md-8">
                <Input
                  type="text"
                  id="current-pair"
                  value={pair}
                  onChange={this.handleOnChangeFor('pair')}
                />
              </div>
            </div>
            <button className="btn btn-primary" onClick={this.calculate}>Calculate</button>
          </div>
          <div className="col-md-6">
            <h5 className="card-title">Result</h5>
            <div className="row">
              <label className="col-md-4">Risk Ratio, %</label>
              <div className="col-md-4">{riskRatio || 'N/A'} %</div>
            </div>
            <div className="row">
              <label className="col-md-4">Lots</label>
              <div className="col-md-4">{(lots && lots.toFixed(3)) || 'N/A'}</div>
            </div>
          </div>
        </div>
      </div>
    )
  }
}

export default PositionCalculator;

