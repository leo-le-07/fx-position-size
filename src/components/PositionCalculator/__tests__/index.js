import React from 'react';
import { shallow } from 'enzyme';
import PositionCalculator from '../index.js';

describe('PositionCalculator', () => {
  test('should match snapshot', () => {
    const wrapper = shallow(<PositionCalculator />);
    expect(wrapper).toMatchSnapshot();
  });
});
