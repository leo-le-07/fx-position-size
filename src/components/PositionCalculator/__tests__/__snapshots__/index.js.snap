// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PositionCalculator should match snapshot 1`] = `
<div
  className="card mb-4"
>
  <div
    className="card-header"
  >
    Position Size Calculator
  </div>
  <div
    className="card-body row"
  >
    <div
      className="col-md-8"
    >
      <h5
        className="card-title"
      >
        Inputs
      </h5>
      <div
        className="form-group row"
      >
        <label
          className="col-md-4"
          htmlFor="account-size"
        >
          Account Size (USD)
        </label>
        <div
          className="col-md-8"
        >
          <UIInput
            className="form-control"
            id="account-size"
            onChange={[Function]}
            type="number"
            value={800}
          />
        </div>
      </div>
      <div
        className="form-group row"
      >
        <label
          className="col-md-4"
          htmlFor="risk-money"
        >
          Risk Money (USD)
        </label>
        <div
          className="col-md-8"
        >
          <UIInput
            className="form-control"
            id="risk-money"
            onChange={[Function]}
            type="number"
            value={20}
          />
        </div>
      </div>
      <div
        className="form-group row"
      >
        <label
          className="col-md-4"
          htmlFor="stop-loss-pips"
        >
          Stop-Loss (pips)
        </label>
        <div
          className="col-md-8"
        >
          <UIInput
            className="form-control"
            id="stop-loss-pips"
            onChange={[Function]}
            type="number"
            value={10}
          />
        </div>
      </div>
      <div
        className="form-group row"
      >
        <label
          className="col-md-4"
          htmlFor="current-pair"
        >
          Current Pair
        </label>
        <div
          className="col-md-8"
        >
          <UIInput
            className="form-control"
            id="current-pair"
            onChange={[Function]}
            type="text"
            value="EURUSD"
          />
        </div>
      </div>
      <button
        className="btn btn-primary"
        onClick={[Function]}
      >
        Calculate
      </button>
    </div>
    <div
      className="col-md-4"
    >
      <h5
        className="card-title"
      >
        Result
      </h5>
      <label>
        Risk Ratio: 
        N/A
         %
      </label>
      <label>
        Lots: 
        N/A
      </label>
    </div>
  </div>
</div>
`;
